lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+       )
+       when module != nil do
+    {arg, state, env} = expand(arg, state, env)
+
+    case arg do
+      keyword when is_list(keyword) ->
+        {nil, make_overridable(state, env, keyword, meta[:context]), env}
+
+      behaviour_module when is_atom(behaviour_module) ->
+        if Code.ensure_loaded?(behaviour_module) and
+             function_exported?(behaviour_module, :behaviour_info, 1) do
+          keyword =
+            behaviour_module.behaviour_info(:callbacks)
+            |> Enum.map(&Introspection.drop_macro_prefix/1)
+
+          {nil, make_overridable(state, env, keyword, meta[:context]), env}
+        else
+          {nil, state, env}
+        end
+
+      _ ->
+        {nil, state, env}
+    end
+  end
+
+  defp expand_macro(
+         meta,
+         Kernel,
+         type,
+         [fields],
+         _callback,
+         state,
+         env = %{module: module}
+       )
+       when type in [:defstruct, :defexception] and module != nil do
+    if Map.has_key?(state.structs, module) do
+      raise ArgumentError,
+            "defstruct has already been called for " <>
+              "#{inspect(module)}, defstruct can only be called once per module"
+    end
+
+    {fields, state, env} = expand(fields, state, env)
+
+    fields =
+      case fields do
+        fs when is_list(fs) ->
+          fs
+
+        _other ->
+          # elixir raises ArgumentError here
+          []
+      end
+
+    fields =
+      fields
+      |> Enum.filter(fn
+        field when is_atom(field) -> true
+        {field, _} when is_atom(field) -> true
+        _ -> false
+      end)
+      |> Enum.map(fn
+        field when is_atom(field) -> {field, nil}
+        {field, value} when is_atom(field) -> {field, value}
+      end)
+
+    state =
+      state
+      |> add_struct_or_exception(env, type, fields, extract_range(meta))
+
+    {{type, meta, [fields]}, state, env}
+  end
+
+  defp expand_macro(
+         meta,
+         Record,
+         call,
+         [_name, _fields] = args,
+         _callback,
+         state,
+         env = %{module: module}
+       )
+       when call in [:defrecord, :defrecordp] and module != nil do
+    range = extract_range(meta)
+    {[name, _fields] = args, state, env} = expand(args, state, env)
+
+    type =
+      case call do
+        :defrecord -> :defmacro
+        :defrecordp -> :defmacrop
+      end
+
+    options = [generated: true]
+
+    state =
+      state
+      |> add_func_to_index(
+        env,
+        name,
+        [{:\\, [], [{:args, [], nil}, []]}],
+        range,
+        type,
+        options
+      )
+      |> add_func_to_index(
+        env,
+        name,
+        [{:record, [], nil}, {:args, [], nil}],
+        range,
+        type,
+        options
+      )
+      |> add_current_env_to_line(meta, env)
+
+    {{{:., meta, [Record, call]}, meta, args}, state, env}
+  end
+
+  defp expand_macro(
+         meta,
+         Kernel,
+         :defprotocol,
```

Comment:
Is there a reason why regular expansion is not enough here? Anything we can do in Elixir to improve it?

Reply:
It won't work without actually evaling the module body

Reply:
Last time we chatted, the theory is that you would be able to expand and track some variables in order to make this work. It would take a bit of complexity but the upside is that it would make the code more generic and make more expansions work. Feel free to ping me about this and we can jump into a chat later?